src/Plutarch/Test/Precompiled.hs

```diff
@@ -0,0 +1,179 @@
+{-# LANGUAGE RankNTypes #-}
+{-# LANGUAGE RecordWildCards #-}
+
+{- | Module: Plutarch.Test.Precompiled
+ Copyright: (C) Liqwid Labs 2022
+ Maintainer: Seungheon Oh <seungheon@mlabs.city>
+ Portability: GHC only
+ Stability: Experimental
+
+ Helpers for building test tree that shares same precompiled script.
+-}
+module Plutarch.Test.Precompiled (
+    Expectation (..),
+    TestCase,
+    (@&),
+    withApplied,
+    TestCompiled (..),
+    testEvalCase,
+    (@>),
+    (@!>),
+    fromPTerm,
+) where
+
+import Acc (Acc)
+import Control.Monad.RWS (MonadReader, MonadWriter, RWS, ask, execRWS, local, tell)
+import Data.Foldable (toList)
+import Data.Tagged (Tagged (Tagged))
+import Data.Text (Text)
+import Plutarch.Evaluate (EvalError, evalScript)
+import Plutarch.Extra.DebuggableScript (DebuggableScript (DebuggableScript, debugScript, script), mustCompileD)
+import Plutarch.Prelude (ClosedTerm, S, Type)
+import PlutusLedgerApi.V1.Scripts (Script, applyArguments)
+import PlutusLedgerApi.V2 (Data)
+import Test.Tasty (testGroup)
+import Test.Tasty.Providers (IsTest, Result, TestTree, run, singleTest, testFailed, testOptions, testPassed)
+import Text.PrettyPrint (
+    Doc,
+    Style,
+    hang,
+    lineLength,
+    renderStyle,
+    style,
+    vcat,
+ )
+import Text.Show.Pretty (ppDoc)
+
+applyDebuggableScript :: DebuggableScript -> [Data] -> DebuggableScript
+applyDebuggableScript (DebuggableScript script debugScript) d =
+    DebuggableScript
+        { script = applyArguments script d
+        , debugScript = applyArguments debugScript d
+        }
+
+-- | @since 0.1.1.0
+data Expectation
+    = Success
+    | Failure
+    deriving stock (Show)
+
+{- | Holds necessary information for each test cases.
+
+ @since 0.1.1.0
+-}
+data TestCase = TestCase
+    { dScript :: DebuggableScript
+    , caseName :: String
+    , expectation :: Expectation
+    }
+
+ourStyle :: Style
+ourStyle = style{lineLength = 80}
+
+ppLogs :: [Text] -> Doc
+ppLogs = \case
+    [] -> "No logs found. Did you forget to build with +development?"
+    logs -> vcat $ ppDoc <$> logs
+
+instance IsTest TestCase where
+    testOptions = Tagged []
+    run _ TestCase{..} _ = return $
+        case (r, expectation) of
+            (Right _, Success) -> testPassed ""
+            (Right x, Failure) -> failWithStyle . unexpectedSuccess x $ dt
+            (Left err, Success) -> failWithStyle . unexpectedFailure err $ dt
+            (Left _, Failure) -> testPassed ""
+      where
+        (r, _, _) = evalScript $ script dScript
+        (_, _, dt) = evalScript $ debugScript dScript
```

Comment:
I'd recommend explicit matches in the body of `run` over `where`-binds like these. They can't have signatures and look _really_ confusing.

Reply:
explicitly matching this would be a nightmare with 6 fields... I'll provide a type signature